src/Common/src/TypeSystem/IL/Stubs/TypeGetTypeMethodThunk.cs

```diff
@@ -0,0 +1,178 @@
+
+using Internal.TypeSystem;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.IL.Stubs
+{
+    /// <summary>
+    /// Helps model the behavior of calls to various Type.GetType overloads. Type.GetType is required to search
+    /// the calling assembly for a matching type if the type name supplied by the user code was not assembly qualified.
+    /// This thunk calls a helper method, passing it a string for what should be considered the "calling assembly".
+    /// </summary>
+    internal class TypeGetTypeMethodThunk : ILStubMethod
+    {
+        private readonly MethodDesc _helperMethod;
+
+        public TypeGetTypeMethodThunk(TypeDesc owningType, MethodSignature signature, MethodDesc helperMethod, string defaultAssemblyName)
+        {
+            OwningType = owningType;
+            Signature = signature;
+            DefaultAssemblyName = defaultAssemblyName;
+
+            _helperMethod = helperMethod;
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return OwningType.Context;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return $"{_helperMethod.Name}_{Signature.Length}_{DefaultAssemblyName}";
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get;
+        }
+
+        public override MethodSignature Signature
+        {
+            get;
+        }
+
+        public string DefaultAssemblyName
+        {
+            get;
+        }
+
+        public override MethodIL EmitIL()
+        {
+            ILEmitter emit = new ILEmitter();
+            ILCodeStream codeStream = emit.NewCodeStream();
+
+            Debug.Assert(Signature[0] == _helperMethod.Signature[0]);
+            codeStream.EmitLdArg(0);
+
+            Debug.Assert(_helperMethod.Signature[1].IsString);
+            codeStream.Emit(ILOpcode.ldstr, emit.NewToken(DefaultAssemblyName));
+
+            for (int i = 2; i < _helperMethod.Signature.Length; i++)
+            {
+                // The helper method could be expecting more arguments than what we have - check for that
+                // The thunk represents one of the 6 possible overloads:
+                // (String), (String, bool), (String, bool, bool)
+                // (String, Func<...>, Func<...>), (String, Func<...>, Func<...>, bool), (String, Func<...>, Func<...>, bool, bool)
+                // We only need 2 helpers to support all 6 overloads. The default value for the bools is false.
+
+                if (i - 1 < Signature.Length)
+                {
+                    // Pass user's parameter
+                    Debug.Assert(_helperMethod.Signature[i] == Signature[i - 1]);
+                    codeStream.EmitLdArg(i - 1);
+                }
+                else
+                {
+                    // Pass a default value
```

Comment:
We are going to have assorted set of methods that need this transform. E.g. GetExecutingAssembly, to just name one example. 

It may be better to make this transform dumb: lookup the method with exact same signature; always append the hidden argument as the last argument and repush the rest; and have the code that fills in the defaults GetType arguments in corelib instead.